### DISCUSSION TITLE
everyday ggplot2 extension at ASA-COWY

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#57] [Show and tell]
Doing a talk on Stat extension and ideas for keeping things light weight this Friday!   Will post slides here later on.

Title:  Everyday ggplot2 extension: New approaches to light-weight 'geom' extension

Abstract: Creating new layers (geom_* functions) in ggplot2 extension has been described as an 'involved process'.  While new educational material has lowered the conceptual barrier to entry, the amount of scaffolding code required for classical extension can make regularly using extension feel prohibitive.  In this talk, I'll discuss strategies and tools to make experimentation with new functionality more light-weight and fun.  A practical application in statistics pedagogy will be presented.

**Comment by EvaMaeRey** [comment]
I'll use that stats pedegogy example for ASA, but maybe gguke in the more fun by hypothetical version of this talk!

**Comment by EvaMaeRey** [comment]
I'm stuck on this GeomArea X ggproto("StatTemp", Stat, compute_panel = my_compute) error at the end of this reprex.  Any insights are welcome!  Combining the stat with GeomPoint and GeomSegment seems to be working, and if I put the computation in compute_group then it also doesn't throw an error with GeomArea, but group-wise compute is not desired.  hmmmm....  

``` r
library(tidyverse)


# Proposed layer composition
compute_dnorm_prop_sds <- function(data, scales, null = .5,
  dist_sds = -4:4){
  
  n <- nrow(data)
  
  sd = sqrt(null * (1 - null)/n) # sd of the null distribution
  
  q <- dist_sds * sd + null
  
  data.frame(x = q + 1) %>%
    mutate(height = dnorm(q, sd = sd, mean = null)) %>%
    mutate(height_max = dnorm(0, sd = sd, mean = 0)) %>%
    mutate(y = .35*n*height/height_max) %>% # This is a bit fragile...
    mutate(xend = x,
           yend = 0)

}  

compute_balance_beam <- function(data, scales, null = .5,
  dist_sds = -4:4){
  
data %>% 
    summarise(xend = max(x),
              x = min(x),
              yend = 0,
              y = 0)

}  

palmerpenguins::penguins %>% 
  mutate(x = species == "Gentoo") %>% 
  compute_dnorm_prop_sds()
#>          x       height height_max           y     xend yend
#> 1 1.392167  0.004964362   14.79855   0.0403897 1.392167    0
#> 2 1.419125  0.164397086   14.79855   1.3375232 1.419125    0
#> 3 1.446084  2.002766503   14.79855  16.2943681 1.446084    0
#> 4 1.473042  8.975776745   14.79855  73.0262914 1.473042    0
#> 5 1.500000 14.798554041   14.79855 120.4000000 1.500000    0
#> 6 1.526958  8.975776745   14.79855  73.0262914 1.526958    0
#> 7 1.553916  2.002766503   14.79855  16.2943681 1.553916    0
#> 8 1.580875  0.164397086   14.79855   1.3375232 1.580875    0
#> 9 1.607833  0.004964362   14.79855   0.0403897 1.607833    0
```

``` r

palmerpenguins::penguins |>
  ggplot() + 
  aes(x = species == "Gentoo") + 
  geom_segment(stat = ggproto("temp", 
                           Stat, 
                           compute_panel = compute_balance_beam))
```

![](https://i.imgur.com/Q7SwEtE.png)

``` r

last_plot() + 
  geom_point(stat = ggproto("temp", 
                           Stat, 
                           compute_panel = compute_dnorm_prop_sds))
```

![](https://i.imgur.com/r8YS6yy.png)

``` r

last_plot() + 
  geom_segment(stat = ggproto("temp", 
                           Stat, 
                           compute_panel = compute_dnorm_prop_sds))
```

![](https://i.imgur.com/30Dnh6X.png)

``` r

last_plot() + 
  geom_area(stat = ggproto("temp", 
                           Stat, 
                           compute_group = compute_dnorm_prop_sds))
```

![](https://i.imgur.com/2jI3o9o.png)

``` r

last_plot() + 
  geom_area(stat = ggproto("temp", 
                           Stat, 
                           compute_panel = compute_dnorm_prop_sds))
#> Error in `geom_area()`:
#> ! Problem while setting up geom.
#> ℹ Error occurred in the 5th layer.
#> Caused by error in `order()`:
#> ! argument 2 is not a vector
```

<sup>Created on 2024-11-06 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
`GeomArea$setup_data()` requires a group column. Your panel computation does not preserve groups, but it should.

**Reply by EvaMaeRey** [reply]
🙌🙏  Thank you... I couldn't see my way out...  That makes sense of course for GeomArea, and I see that need in setup_data.  😮‍💨

**Reply by EvaMaeRey** [reply]
Same questioned layer as in #39 ... 🤷‍♀️

**Reply by EvaMaeRey** [reply]
I guess I'd seen the discrete scales flexibility, but it didn't dawn on me until @teunbrand walked through the notes at the release party that it would be really useful for the prop case `[scale_x_discrete](https://ggplot2.tidyverse.org/reference/scale_discrete.html)(
    palette = scales::[pal_manual](https://scales.r-lib.org/reference/pal_manual.html)([c](https://rdrr.io/r/base/c.html)(1:3, 5:7))
  )` so you aren't stuck with c(1,2) as the numerics underlying you what are really F, T categories - and 0 and 1 are the numeric you really want.  I'm just getting a feel for it - and seems to very work well in this prop.test viz - letting me delete a bunch of awkward work arounds in compute (subtracting 1 in a bunch of places).   🎉📊 yay .  :-)

**Comment by EvaMaeRey** [comment]
working slides.  Event postponed, so there is definitely a roughness to these still.   https://evamaerey.github.io/mytidytuesday/2024-10-29-asa-cowy-fall-2024/asa-cowy-fall-2024#1

**Reply by EvaMaeRey** [reply]
Now public and updating gguke, to balance out what feels like too much statistics extension focus... https://github.com/EvaMaeRey/gguke  moving from megafunction to composable w/ approach similar to statRegion in #10:  Stat left joins on positions (fret, finger), and input column is the name of places (chord names).

**Comment by EvaMaeRey** [comment]
New favorite working analogy for talk: Stats instructors as longform graphical poets... 🤯😅

**Reply by EvaMaeRey** [reply]
And they are milhojas cake bakers!  (vs. a couple of layers)  https://www.tangopastry.com/menu/p/mil-hojas-cake

![image](https://github.com/user-attachments/assets/2678099b-900d-4398-803f-a4079fd2ea37)

**Comment by EvaMaeRey** [comment]
The coding approaches to showing the visuals are very different in these two 'test as linear regression' resources: r: show all your work - w/ ggplot2; python: wrap up the plotting - I think w/ matplotlib

r: https://lindeloev.github.io/tests-as-linear
python: https://www.georgeho.org/tests-as-linear/

So much 😍 for connecting stats tests: https://twitter.com/jonaslindeloev/status/1110907133833502721
